synapse/rest/media/v1/media_storage.py

```diff
@@ -143,12 +143,9 @@ async def fetch_media(self, file_info: FileInfo) -> Optional[Responder]:
         """
 
         path = self._file_info_to_path(file_info)
-        local_path = os.path.join(self.local_media_directory, path)
-        if os.path.exists(local_path):
-            return FileResponder(open(local_path, "rb"))
```

Comment:
this is just moved down by a few lines.

Reply:
In general I wonder if a bunch of this code could be de-duplicated and simplified by generating a set of paths to check (the "correct" path and maybe the legacy one) something like:

```python
paths = [local_path]
if legacy_path:
    paths.append(legacy_path)

for path in paths:
    # Check if exists in local media dir

for path in paths:
    for provider in self.storage_providers:
        # Check if can fetch from provider
```

🤷 There's potentially a couple ways to organize those loops and such too.

Overall this seems like a fine change though I do wonder if it makes sense to check each provider for the "correct" file name first, then the legacy one or to iterate through each provider (as the PR has now). I think we only expect one match ever, so it shouldn't really matter?

Reply:
>  I do wonder if it makes sense to check each provider for the "correct" file name first, then the legacy one or to iterate through each provider (as the PR has now). 

The current implementation means that we can check all local stores first, before making any S3 requests. Flipping the order around would mean adding the delay of an S3 lookup for a file that is stored in the local system with a legacy name.

Reply:
I've deduplicated the code a bit.

Reply:
I find the logic much easier to follow now! Thanks. 👍